### PR TITLE
Make pallet Assets instantiable

### PR DESCRIPTION
**Pull request by olanod**
As mentioned in https://github.com/paritytech/substrate/issues/8453#issuecomment-809272236, making the assets pallet instantiable allows one to better group sets of assets that share something in common instead of putting everything in a single big group that might require some sort of filtering on the pallets that depend on some fungibles.

### CODE REVIEW

**Comment by shawntabrizi** [comment]
@olanod asset creation is permissionless, so i don't see how such organization you are striving for will be enforced.

Do you have a more concrete idea on how you want to use this stuff?

**Comment by olanod** [comment]
But it also can be permissioned only right? At least some of the assets in our chain will be created only by an authority i.e. a set of valid collaterals for a given stable coin. Also I'm looking into assets pallet mostly as an implementor of the `fungibles::*` traits and not much for its own dispatchable calls.

**Comment by shawntabrizi** [comment]
@olanod no, it is current not possible to make it permissioned without some sort of wrapper on what functions can be called:

https://github.com/paritytech/substrate/blob/master/frame/assets/src/lib.rs#L368

**Comment by olanod** [comment]
@shawntabrizi but that's fine no? I can have that wrapper as an extra pallet or a proxy that filters calls and still benefit from the instantiability of Assets pallet. I would say it's even more compelling to have different instances in this case, say I have one instance with some access control filter for cool people only and another one that is general purpose.

**Comment by shawntabrizi** [comment]
@olanod please take a look at the diff that github provides. We do not use spaces in Substrate, so please  update them to tabs.

**Comment by olanod** [comment]
My bad ... will have to disable the format on save 😅

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
@thiolliere can you take a quick look here?

**Comment by olanod** [comment]
The easiest way I found to revert the spaces formatting was to format again overriding the `hard_tabs` setting of rustfmt, rustmt has a lot of opinions on how the code should look like but as its not the scope of the PR I tried to leave out as much of the extra formatting that was not directly related to the feature as possible splitting chunks as far as `git add --patch --interactive` allowed me but there might be a bit of formatted code left here and there that I hope you wont mind(and might be for the best).

File: frame/assets/src/lib.rs [hunk 1]
```diff
@@ -473,25 +483,37 @@ pub mod pallet {
...
 				}
-				ensure!(details.accounts == witness.accounts, Error::<T>::BadWitness);
-				ensure!(details.sufficients == witness.sufficients, Error::<T>::BadWitness);
-				ensure!(details.approvals == witness.approvals, Error::<T>::BadWitness);
-
-				for (who, v) in Account::<T>::drain_prefix(id) {
+				ensure!(
+					details.accounts == witness.accounts,
+					Error::<T, I>::BadWitness
+				);
```
Comment (thiolliere) [inline frame/assets/src/lib.rs:494]: in substrate the line length is 100 character, those ensure doesn't need to go to newline.

File: frame/assets/src/lib.rs [hunk 2]
```diff
@@ -881,14 +911,23 @@ pub mod pallet {
 		) -> DispatchResult {
 			let origin = ensure_signed(origin)?;
 
-			ensure!(name.len() <= T::StringLimit::get() as usize, Error::<T>::BadMetadata);
-			ensure!(symbol.len() <= T::StringLimit::get() as usize, Error::<T>::BadMetadata);
+			ensure!(
+				name.len() <= T::StringLimit::get() as usize,
+				Error::<T, I>::BadMetadata
+			);
```
Comment (thiolliere) [inline frame/assets/src/lib.rs:917]: same no need for new line

File: frame/assets/src/lib.rs [hunk 3]
```diff
@@ -968,11 +1007,17 @@ pub mod pallet {
...
 
-			ensure!(name.len() <= T::StringLimit::get() as usize, Error::<T>::BadMetadata);
-			ensure!(symbol.len() <= T::StringLimit::get() as usize, Error::<T>::BadMetadata);
-
-			ensure!(Asset::<T>::contains_key(id), Error::<T>::Unknown);
-			Metadata::<T>::try_mutate_exists(id, |metadata| {
+			ensure!(
+				name.len() <= T::StringLimit::get() as usize,
+				Error::<T, I>::BadMetadata
+			);
```
Comment (thiolliere) [inline frame/assets/src/lib.rs:1013]: same

**Review by thiolliere — APPROVED** [review]

**Review by thiolliere — CHANGES_REQUESTED** [review]
ah I forgot, benchmarks need to be updated too, right now benchmark are only generated for default instance not for other ones.
```diff
diff --git a/frame/assets/src/benchmarking.rs b/frame/assets/src/benchmarking.rs
index 227d45623d..d02fe7b74b 100644
--- a/frame/assets/src/benchmarking.rs
+++ b/frame/assets/src/benchmarking.rs
@@ -135,7 +135,7 @@ fn assert_event<T: Config>(generic_event: <T as Config>::Event) {
        }));
 }
 
-benchmarks! {
+benchmarks_instance_pallet! {
        create {
                let caller: T::AccountId = whitelisted_caller();
                let caller_lookup = T::Lookup::unlookup(caller.clone());
```

But there will be other changes inside the file

**Comment by shawntabrizi** [comment]
@thiolliere check again?

**Review by thiolliere — APPROVED** [review]
looks good to me

**Comment by thiolliere** [comment]
bot merge

**Comment by olanod** [comment]
Awesome! thanks @shawntabrizi was delaying that for the weekend 👍🏽

**Pull request merged**